parlai/core/torch_agent.py

```diff
@@ -1551,30 +1524,93 @@ def match_batch(self, batch_reply, valid_inds, output=None):
                 batch_reply[i]['text_candidates'] = cands
         return batch_reply
 
-    def last_reply(self, use_reply='label'):
+    def observe(self, observation):
         """
-        Retrieve the last reply from the model.
+        Process incoming message in preparation for producing a response.
 
-        If available, we use the true label instead of the model's prediction.
+        This includes remembering the past history of the conversation.
+        """
+        # TODO: Migration plan: TorchAgent currently supports being passed
+        # observations as vanilla dicts for legacy interop; eventually we
+        # want to remove this behavior and demand that teachers return Messages
+        observation = Message(observation)
 
-        By default, batch_act stores the batch of replies and this method
-        will extract the reply of the current instance from the batch.
+        if self.__expecting_to_reply:
+            raise RuntimeError(
+                "Last observe() had a label, but no call to self_observe ever "
+                "happened. You are likely making multiple observe() calls without "
+                "a corresponding act(). This was changed in #2043. File a GitHub "
+                "issue if you require assistance."
+            )
+
+        if self.__expecting_clear_history:
+            raise RuntimeError(
+                "Last observe() was episode_done, but we never saw a corresponding "
+                "self_observe to clear the history, probably because you missed an "
+                "act(). This was changed in #2043. File a GitHub issue if you require "
+                "assistance."
+            )
+
+        if observation.get('episode_done'):
+            self.__expecting_clear_history = True
+        elif 'labels' in observation or 'eval_labels' in observation:
+            # make sure we note that we're expecting a reply in the future
+            self.__expecting_to_reply = True
```

Comment:
so this won't catch errors in interactive mode?

Reply:
The old case of interactive bugs isn’t possible because agents never use shared state to handle reply handling now.

Reply:
Unless you’re saying we wouldn’t catch a missing call to self observe in interactive, in which case I think you’re right.

Reply:
I think my concern about an assert without a label is we couldn’t handle multiparty correctly then.